vm/src/stdlib/itertools.rs

```diff
@@ -4,14 +4,74 @@ use std::ops::{AddAssign, SubAssign};
 
 use num_bigint::BigInt;
 
-use crate::function::OptionalArg;
+use crate::function::{OptionalArg, PyFuncArgs};
 use crate::obj::objbool;
 use crate::obj::objint::{PyInt, PyIntRef};
 use crate::obj::objiter::{call_next, get_iter, new_stop_iteration};
+use crate::obj::objtype;
 use crate::obj::objtype::PyClassRef;
 use crate::pyobject::{PyClassImpl, PyObjectRef, PyRef, PyResult, PyValue};
 use crate::vm::VirtualMachine;
 
+#[pyclass(name = "chain")]
+#[derive(Debug)]
+struct PyItertoolsChain {
+    iterables: Vec<PyObjectRef>,
+    cur: RefCell<usize>,
+    cur_iter: RefCell<Option<PyObjectRef>>,
+}
+
+impl PyValue for PyItertoolsChain {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("itertools", "chain")
+    }
+}
+
+#[pyimpl]
+impl PyItertoolsChain {
+    #[pymethod(name = "__new__")]
+    fn new(_cls: PyClassRef, args: PyFuncArgs, vm: &VirtualMachine) -> PyResult {
+        Ok(PyItertoolsChain {
+            iterables: args.args,
+            cur: RefCell::new(0),
+            cur_iter: RefCell::new(None),
+        }
+        .into_ref(vm)
+        .into_object())
+    }
+
+    #[pymethod(name = "__next__")]
+    fn next(&self, vm: &VirtualMachine) -> PyResult {
+        while *self.cur.borrow() < self.iterables.len() {
+            if self.cur_iter.borrow().is_none() {
+                *self.cur_iter.borrow_mut() =
+                    Some(get_iter(vm, &self.iterables[*self.cur.borrow()])?);
+            }
+
+            // can't be directly inside the 'match' clause, otherwise the borrows collide.
+            let obj = call_next(vm, self.cur_iter.borrow().as_ref().unwrap());
+            match obj {
+                Ok(ok) => return Ok(ok),
+                Err(err) => {
+                    if objtype::isinstance(&err, &vm.ctx.exceptions.stop_iteration) {
+                        *self.cur.borrow_mut() += 1;
+                        *self.cur_iter.borrow_mut() = None;
+                    } else {
+                        return Err(err);
+                    }
+                }
+            }
+        }
```

Comment:
I think this could be represented better with recursion than a loop - maybe change the `while` to an `if` with the `new_stop_iteration` in the `else`, and then after incrementing `self.cur`, call `self.next(vm)`. That will also allow you to use `obj.or_else(|err| { ... })` instead of the `match` statement.

Reply:
Hmm, okay. I based this function on CPython, which has a while loop itself. But I guess the likeliness of an semi-infinite recursion here is very low.

Reply:
Actually, the recursion trick doesn't work after I applied your suggestions in https://github.com/RustPython/RustPython/pull/986#discussion_r285356963 :(

Reply:
(Because the mutable borrow persists across calls)

Reply:
Ah, okay. This seems fine then.